### ISSUE TITLE
Travis-CI + Github API limit

### ISSUE THREAD

**Issue by fprochazka** [composer/composer#1314]
Is there any suggested way to solve the Github API limit problem? Or will my builds just randomly fail?

https://travis-ci.org/Kdyby/Framework/jobs/3142911

Not sure, if this is problem of Composer or Travis-CI.

**Comment by aertmann** [comment]
Had this issue as well when using ``create-project``. Neither adding ``--no-interaction`` nor ``--prefer-source`` solved the problem, but using ``git clone`` and ``composer install`` didn't have the issue. No idea why there would be any difference, but seems there is. Maybe it's handled differently internally and those arguments are only used for the root repository. Just a tip if anyone else comes across the same issue.

**Comment by bojanz** [comment]
http://blog.wyrihaximus.net/2015/09/github-auth-token-on-travis/ documents most of the solution.

What we did:
1) Created the token. No scopes (read only access to public info only).
2) Added it to Travis as the GITHUB_OAUTH_TOKEN env variable.
3) Added git config --global github.accesstoken $GITHUB_OAUTH_TOKEN to .travis.yml
The article advocates doing composer config github-oauth.github.com instead, but that still failed on --prefer-dist for us. Composer bug?
4) Removed --prefer source, kept -n.

**Comment by mykz** [comment]
@bojanz really nice solution, Thanks for taking time to post it. :+1:

**Comment by danepowell** [comment]
Hmm... so Github was supposed to have fixed this by removing anonymous rate limits:
https://github.com/composer/composer/issues/4884#issuecomment-195229989

But I'm still seeing composer installs failing on Travis with errors like this:
"Downloading: Connecting... Failed to download composer/installers from dist: Could not authenticate against github.com"

We never added an OAuth token to composer.json or .travis.yml, so it shouldn't be authenticating. I think this must be due to rate limiting.

**Comment by bojanz** [comment]
Yeah, I don't think there's a way to avoid rate limiting in Travis, cause it's simply a case of many builds coming from the same IPs. No matter how high the limit is on GitHub, Travis will reach it.

**Comment by danepowell** [comment]
Actually it seems to be only intermittently failing. I wonder if some of the Travis IPs are still blacklisted from the old rate limit and it just takes a while to clear them or something...

**Comment by hkdobrev** [comment]
Travis CI uses a lot of different IPs and it changes them on every build. That's why it cannot provide a list or a range to whitelist in firewalls. I guess GitHub has rate-limited some IP, but the next built uses another one.

**Comment by alexislefebvre** [comment]
I have successfully tested a workaround which consist in caching the `vendor/` directory, an archive is created with `tar` and stored in Travis CI's cache. I presented this method on [Stack Overflow](http://stackoverflow.com/a/36085819/2257664). A `composer update --prefer-source` takes 30 seconds when this cache is used, and 2 minutes without this cache.

**Comment by spekary** [comment]
I recently ran into this problem, grunted and groaned over it for a couple of days, and wanted to post a fix for our specific problem as it might help others. The cause of this for us was that we had "repository" entries in our composer.json file pointing to our github source for some new repositories under development. Even though we had registered these with packagist, we were still seeing the problem. It wasn't until we removed these repository entries from the composer.json file so that packagist could be the pointer to them, did we finally see our travis builds start to work consistently.
